File: Data Structures & Algorithms/max-water-container/submission-2.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int l = 0, r = heights.size() - 1;
        int max_volume = 0;
        while (l < r) {
            int volume = (r - l) * min(heights[l], heights[r]);
            max_volume = max(volume, max_volume);
            
            if (heights[l] > heights[r]) {
                r --;
            } else {
                l ++;
            }
        }
        return max_volume;
    }
};

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-2.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        Deque<Integer> stack = new ArrayDeque<>();
        int maxArea = 0;

        // i == heights.length 时，加入一个虚拟的高度 0
        for (int i = 0; i <= heights.length; i++) {
            int currentHeight =
                    i == heights.length ? 0 : heights[i];

            while (!stack.isEmpty()
                    && heights[stack.peek()] > currentHeight) {

                int h = heights[stack.pop()];

                int width = stack.isEmpty()
                        ? i
                        : i - stack.peek() - 1;

                maxArea = Math.max(maxArea, h * width);
            }

            stack.push(i);
        }

        return maxArea;
    }
}
